lib/connections/quic_listen.go

```diff
@@ -101,57 +105,32 @@ func (t *quicListener) serve(stop chan struct{}) error {
 		l.Infoln("Listen (BEP/quic):", err)
 		return err
 	}
+	defer listener.Close()
 
 	l.Infof("QUIC listener (%v) starting", packetConn.LocalAddr())
 	defer l.Infof("QUIC listener (%v) shutting down", packetConn.LocalAddr())
 
-	// Accept is forever, so handle stops externally.
-	go func() {
-		select {
-		case <-stop:
-			_ = listener.Close()
-		}
-	}()
-
 	for {
-		// Blocks forever, see https://github.com/lucas-clemente/quic-go/issues/1915
-		session, err := listener.Accept()
-
 		select {
-		case <-stop:
-			if err == nil {
-				_ = session.Close()
-			}
+		case <-ctx.Done():
 			return nil
 		default:
 		}
+
+		session, err := listener.Accept(ctx)
 		if err != nil {
 			if err, ok := err.(net.Error); !ok || !err.Timeout() {
```

Comment:
Also don't warn for `err == context.Canceled`. And in principle that could be used to break the loop instead of the select above (but I think the explicit check at the beginning like now is cleaner).

Reply:
Fixd. I'm keeping the select because there is no guarantee that the QUIC package won't do a `return fmt.Errorf("accept: %v", err)` itself tomorrow and then the `==` will fail. (Right now it works.) (Longer term we should use `xerrors.Is`...)